site/content/features/stringtemplates.md

```diff
@@ -0,0 +1,466 @@
+
+I don't think that many programmers will create `StringTemplate` instances, but there are three ways to do so:
+
+* With the `RAW` processor:
+          ```
+RAW."Hello, \{name}!"
+```
+
+
+* With the static `of` methods:
+          ```
+StringTemplate.of(List("Hello, ", "!"), List(name))
+StringTemplate.of("Hello, Sailor!")
+```
+
+
+* With the static `combine` methods:
+          ```
+StringTemplate.combine(RAW."Hello, \{name}! ", RAW."Next year, you'll be \{age + 1}.")
+```
+
+
+
+There are two `combine` methods, with parameter types `StringTemplate...` and `List<StringTemplate>`.
+
+Finally, the `process` method lets you apply a processor, just in case you prefer
+
+```
+template.process(processor)
+```
+
+over
+
+```
+processor.process(template)
+```
+
+This sandbox shows the API in action, even though you probably only need to worry about the first two methods. 
+
+{{< sandbox version=java19 preview="true" mainclass="StringTemplateDemo" >}}{{< sandboxsource "StringTemplateDemo.java" >}}
+import static java.lang.StringTemplate.RAW;
+import java.util.List;
+
+public class StringTemplateDemo {
+   public static void main(String[] args) {
+      String name = "Fred";
+      int age = 42;
+      StringTemplate template = RAW."Hello, \{name}! Next year, you'll be \{age + 1}.";
+      System.out.println(template.fragments());
+      System.out.println(template.values());
+
+      // Two interpolate methods: one static, one instance
+      System.out.println(StringTemplate.interpolate(
+         template.fragments().stream().map(String::toUpperCase).toList(), template.values()));
+      System.out.println(StringTemplate.of(
+         template.fragments().stream().map(String::toUpperCase).toList(),
+            template.values()).interpolate());
+
+      // Demo of the of and combine methods.
+      // For each entry in System.getProperties(), one StringTemplate is created.
+      // They are all combined and then processed with the STR processor
+      template = StringTemplate.combine(
+         System.getProperties()
+            .entrySet()
+            .stream()
+            .map(e -> StringTemplate.of(List.of(e.getKey() + ": ", "\n"),
+               List.of(e.getValue())))
+            .toList());
+      System.out.println(template.process(STR));
+   }
+}
+{{< /sandboxsource >}}
+{{< /sandbox >}}
+
+## Writing Your Own Template Processor
+
+The `format` method of a template processor turns `StringTemplate` instances into objects. Since `StringTemplate.Processor` is a functional interface, you can construct an instance from a lambda expression. Here is a simple example that places the values in boxes:
+
+{{< sandbox version=java19 preview="true" mainclass="MyFirstTemplateProcessor" >}}{{< sandboxsource "MyFirstTemplateProcessor.java" >}}
+public class MyFirstTemplateProcessor {
```

Comment:
This example has a couple if compilation errors and also leads to an `IndexOutOfBoundsException`. Here is how I fixed it:

```
public class MyFirstTemplateProcessor {
   public static StringTemplate.Processor<String, RuntimeException> BOX =
      template -> {
         StringBuilder result = new StringBuilder();
         for (int i = 0; i < template.fragments().size(); i++) {
            if (i > 0) {
               result.append('[');
               result.append("" + template.values().get(i - 1));
               result.append(']'); 
            }
            result.append(template.fragments().get(i));
         }
         return result.toString();
      };

   public static void main(String[] args) {
      var name="Fred";
      var age = 42;
      System.out.println(BOX."Hello, \{name}! Next year, you'll be \{age + 1}.");
   }
}
```